readme.md
<!--
  Created at: 08/13/2021 02:13:50 Friday
  Modified at: 08/13/2021 02:42:17 AM Friday
-->

# numToWord

This lib converts the number (currently just `int`) to words (one, two...)

The implementation is currently just in english, but more will be added
(need refactor a little bit)

---

## Usage

```nim
import numToWord

doAssert 1.toWord == "one"
doAssert 123_456.toWord == "one hundred, twenty three thousand, four hundred, fifty six"
```

---

## Installation

Minimum [Nim](https://nim-lang.org) version is `1.2.2`

Please choice one installation method:

- Automatically with nimble
  <!-- ```bash
  nimble install numToWord
  ```
  or -->
  ```bash
  nimble install https://github.com/thisago/numToWord
  ```
Or
- Manually
  ```bash
  git clone https://github.com/thisago/numToWord
  cd numToWord/
  nimble install
  ```

## Docs

A online docs is hosted in [Github Pages](https://thisago.github.io/numToWord/numToWord.html)

---

## License

MIT


changelog.md
<!--
  Created at: 08/13/2021 02:14:48 Friday
  Modified at: 08/13/2021 02:18:30 AM Friday
-->

# Changelog

Verbose log of changes

---

## Version 0.1.0 (08/13/2021 02:15:15 Friday)

- Added a improved of [this JS implementation](https://github.com/marlun78/number-to-words/blob/master/src/toWords.js)


tests/test1.nim
#[
  Created at: 08/12/2021 15:00:22 Thursday
  Modified at: 08/12/2021 07:28:19 PM Thursday
]#

import std/unittest
import numWord

suite "Number to words":
  test "0-9":
    check 0.toWords == "zero"
    check 1.toWords == "one"
    check 2.toWords == "two"
    check 3.toWords == "three"
    check 4.toWords == "four"
    check 5.toWords == "five"
    check 6.toWords == "six"
    check 7.toWords == "seven"
    check 8.toWords == "eight"
    check 9.toWords == "nine"
  test "10-19":
    check 10.toWords == "ten"
    check 11.toWords == "eleven"
    check 12.toWords == "twelve"
    check 13.toWords == "thirteen"
    check 14.toWords == "fourteen"
    check 15.toWords == "fifteen"
    check 16.toWords == "sixteen"
    check 17.toWords == "seventeen"
    check 18.toWords == "eighteen"
    check 19.toWords == "nineteen"
  test "20-29":
    check 20.toWords == "twenty"
    check 21.toWords == "twenty one"
    check 22.toWords == "twenty two"
    check 23.toWords == "twenty three"
    check 24.toWords == "twenty four"
    check 25.toWords == "twenty five"
    check 26.toWords == "twenty six"
    check 27.toWords == "twenty seven"
    check 28.toWords == "twenty eight"
    check 29.toWords == "twenty nine"


src/numToWord.nim
#[
  Created at: 08/12/2021 15:04:05 Thursday
  Modified at: 08/13/2021 02:41:05 AM Friday
]#

##[
  num Word
  --------

  Converts numbers to words

  Improved implementation of https://github.com/marlun78/number-to-words/blob/master/src/toWords.js
]##

{.experimental: "codeReordering".}

from std/strutils import repeat, strip, parseInt, join, replace
from std/math import floor

type
  Values = enum
    Hundred, Thousand, Million, Billion, Trillion, Quadrillion

const
  ten = 10
  max = 9999999999999999.int
  values = [
    100,
    1000,
    1000000,
    1000000000,
    1000000000000.int,
    1000000000000000.int,
  ]

const
  # lessThanTwenty = ["zero", "um", "dois", "tres", "quatro", "cinco", "seis", "sete", "oito", "nove", "dez","onze", "doze", "treze", "quatorze", "quinze", "dezesseis", "dezessete", "dezoito", "dezenove"]
  # tenthsLessThanHundred = ["zero", "dez", "vinte", "trinta", "quarenta", "cinquenta", "sessenta", "setenta", "oitenta", "noventa"]
  lessThanTwenty = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten","eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
  tenthsLessThanHundred = ["zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
  moreThanTenths = ["hundred", "thousand", "million", "billion", "trillion", "quadrillion"]
  minus = "minus"

proc toWord*(number: int): string =
  ## Converts the given `number` to string
  ## TODO: Ordinal implementation like JS one
  if number > max:
    return $number
  result = generateWords(number)

proc generateWords(number: int; words = newSeq[string]()): string =
  var
    remainder: int
    word: string
    number = number
    words = words
  if number == 0:
    result = if words.len == 0: lessThanTwenty[0] else: words.join(" ")
    if result[^1] == ',':
      result = result[0..^1]
    return result
  if number < 0:
    words.add minus
    number = abs(number)
  if number < 20:
    remainder = 0
    word = lessThanTwenty[number]
  elif number < values[ord(Hundred)]:
    remainder = number mod ten
    word = tenthsLessThanHundred[floor(number / ten).int]
    if remainder != 0:
      word.add " " & lessThanTwenty[remainder]
      remainder = 0
  else:
    for i in 0..<values.len:
      let value = values[i]
      if number < values[i + 1]:
        remainder = number mod value
        word = generateWords(floor(number / value).int) & " " & moreThanTenths[i] & ","
        break

  words.add word
  return generateWords(remainder, words)


when isMainModule:
  echo 0.toWord
  echo 1.toWord
  echo 10.toWord
  echo 14.toWord
  echo 21.toWord
  echo 51.toWord
  echo 58.toWord
  echo 101.toWord
  echo 131.toWord
  echo 1131.toWord
  echo 123456.toWord
  echo 114564531.toWord


src/numWord.nim
#[
  Created at: 08/12/2021 15:04:05 Thursday
  Modified at: 08/12/2021 07:27:59 PM Thursday
]#

##[
  num Word
  --------

  Converts numbers to words

  Implementation of https://github.com/marlun78/number-to-words/blob/master/src/toWords.js
]##

{.experimental: "codeReordering".}

from std/strutils import repeat, strip, parseInt, join, replace
from std/math import floor

type
  Values = enum
    Hundred, Thousand, Million, Billion, Trillion, Quadrillion

const
  ten = 10
  max = 9999999999999999.int
  values = [
    100,
    1000,
    1000000,
    1000000000,
    1000000000000.int,
    1000000000000000.int,
  ]

const
  # lessThanTwenty = ["zero", "um", "dois", "tres", "quatro", "cinco", "seis", "sete", "oito", "nove", "dez","onze", "doze", "treze", "quatorze", "quinze", "dezesseis", "dezessete", "dezoito", "dezenove"]
  # tenthsLessThanHundred = ["zero", "dez", "vinte", "trinta", "quarenta", "cinquenta", "sessenta", "setenta", "oitenta", "noventa"]
  lessThanTwenty = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten","eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
  tenthsLessThanHundred = ["zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
  moreThanTenths = ["hundred", "thousand", "million", "billion", "trillion", "quadrillion"]

proc toWords*(number: int; asOrdinal = false): string =
  var words: seq[string]

  if number > max:
    return $number

  result = generateWords(number)
  # return asOrdinal ? makeOrdinal(words) : words

proc generateWords(number: int; words = newSeq[string]()): string =
  var
    remainder: int
    word: string
    number = number
    words = words

  if number == 0:
    result = if words.len == 0: "zero" else: words.join(" ")
    if result[^1] == ',':
      result = result[0..^1]
    return result
  if number < 0:
    words.add("minus")
    number = abs(number)

  if number < 20:
    remainder = 0
    word = lessThanTwenty[number]

  elif number < values[ord(Hundred)]:
    remainder = number mod ten
    word = tenthsLessThanHundred[floor(number / ten).int]
    if remainder != 0:
      word.add " " & lessThanTwenty[remainder]
      remainder = 0
  else:
    for i in 0..<values.len:
      let value = values[i]
      if number < values[i + 1]:
        remainder = number mod value
        word = generateWords(floor(number / value).int) & " " & moreThanTenths[i + 1] & ","
        break

  words.add word
  return generateWords(remainder, words)


when isMainModule:
  echo 1.toWords
  echo 10.toWords
  echo 14.toWords
  echo 21.toWords
  echo 51.toWords
  echo 58.toWords
  echo 101.toWords
  echo 131.toWords
  echo 1131.toWords
